File: c_src/tidesdb.c
#include <erl_nif.h>
#include <tidesdb/db.h>
#include <stdint.h>
#include "dynamic_array.h"
#include "atoms.h"
#include "resources.h"
#include "errors.h"
#include "db.h"
#include "cf.h"
#include "txn.h"
#include "iter.h"
#include "etdb_compound.h"

static int load(ErlNifEnv *env, void **priv_data, ERL_NIF_TERM load_info) {
  if (etdb_make_resources(env) != 0)
    return 1;

  etdb_make_atoms(env);
  etdb_make_errors(env);

  return 0;
}

static ErlNifFunc nif_funcs[] = {
  /* tdb functions */
  {"open", 2, etdb_open, ERL_NIF_DIRTY_JOB_IO_BOUND},
  {"close", 1, etdb_close, ERL_NIF_DIRTY_JOB_IO_BOUND},

  {"create_column_family", 3, etdb_create_column_family, ERL_NIF_DIRTY_JOB_IO_BOUND},
  {"get_column_family", 2, etdb_get_column_family, ERL_NIF_DIRTY_JOB_IO_BOUND},
  {"list_column_families", 1, etdb_list_column_families, ERL_NIF_DIRTY_JOB_IO_BOUND},

  {"txn_begin", 1, etdb_txn_begin, ERL_NIF_DIRTY_JOB_IO_BOUND},
  {"txn_put", 4, etdb_txn_put, ERL_NIF_DIRTY_JOB_IO_BOUND},
  {"txn_get", 3, etdb_txn_get, ERL_NIF_DIRTY_JOB_IO_BOUND},
  {"txn_delete", 3, etdb_txn_delete, ERL_NIF_DIRTY_JOB_IO_BOUND},
  {"txn_rollback", 1, etdb_txn_rollback, ERL_NIF_DIRTY_JOB_IO_BOUND},
  {"txn_commit", 1, etdb_txn_commit, ERL_NIF_DIRTY_JOB_IO_BOUND},

  {"iter_new", 2, etdb_iter_new},
  {"iter_close", 1, etdb_iter_close},
  {"iter_valid", 1, etdb_iter_valid},
  {"iter_seek_to_first", 1, etdb_iter_seek_to_first},
  {"iter_seek_to_last", 1, etdb_iter_seek_to_last},
  {"iter_key", 1, etdb_iter_key},
  {"iter_value", 1, etdb_iter_value},
  {"iter_next", 1, etdb_iter_next},
  {"iter_prev", 1, etdb_iter_prev},

  /* compound multi-operation functions, to avoid jumping back-and-forth between schedulers */
  {"scan", 3, etdb_scan, ERL_NIF_DIRTY_JOB_IO_BOUND},
};

ERL_NIF_INIT(Elixir.TidesDB.NIF, nif_funcs, load, NULL, NULL, NULL)

File: lib/tides_db.ex
defmodule TidesDB do
  @moduledoc """
  Documentation for `TidesDB`.
  """

  @doc """
  """

  alias TidesDB.NIF
  alias TidesDB.DB
  alias TidesDB.ColumnFamily
  alias TidesDB.Transaction
  alias TidesDB.Iterator

  def open(path, config \\ []) when is_binary(path) and path != "" and is_list(config) do
    name = String.to_charlist(path)

    case NIF.open(name, config) do
      {:ok, ref} ->
        {:ok, %DB{ref: ref}}

      error ->
        error
    end
  end

  def close(%DB{} = db) do
    NIF.close(db.ref)
  end

  def create_column_family(%DB{} = db, name, config \\ []) when is_binary(name) and name != "" and is_list(config) do
    name = String.to_charlist(name)

    NIF.create_column_family(db.ref, name, config)
  end

  def get_column_family(%DB{} = db, name) when is_binary(name) and name != "" do
    name = String.to_charlist(name)

    case NIF.get_column_family(db.ref, name) do
      {:ok, ref} ->
        {:ok, %ColumnFamily{ref: ref}}

      error ->
        error
    end
  end

  def list_column_families(%DB{} = db) do
    case NIF.list_column_families(db.ref) do
      {:ok, names} ->
        {:ok, Enum.map(names, &:erlang.list_to_binary/1)}

      error ->
        error
    end
  end

  def txn_begin(%DB{} = db) do
    case NIF.txn_begin(db.ref) do
      {:ok, ref} ->
        {:ok, %Transaction{ref: ref}}

      error ->
        error
    end
  end

  def txn_put(%Transaction{} = txn, %ColumnFamily{} = cf, key, value) when is_binary(key) and is_binary(value) do
    NIF.txn_put(txn.ref, cf.ref, key, value)
  end

  def txn_get(%Transaction{} = txn, %ColumnFamily{} = cf, key) when is_binary(key) do
    NIF.txn_get(txn.ref, cf.ref, key)
  end

  def txn_delete(%Transaction{} = txn, %ColumnFamily{} = cf, key) when is_binary(key) do
    NIF.txn_delete(txn.ref, cf.ref, key)
  end

  def txn_rollback(%Transaction{} = txn) do
    NIF.txn_rollback(txn.ref)
  end

  def txn_commit(%Transaction{} = txn) do
    NIF.txn_commit(txn.ref)
  end

  def iter_new(%Transaction{} = txn, %ColumnFamily{} = cf) do
    case NIF.iter_new(txn.ref, cf.ref) do
      {:ok, ref} ->
        {:ok, %Iterator{ref: ref}}

      error ->
        error
    end
  end

  def iter_close(%Iterator{} = iter) do
    NIF.iter_close(iter.ref)
  end

  def iter_valid?(%Iterator{} = iter) do
    NIF.iter_valid(iter.ref)
  end

  def iter_seek_to_first(%Iterator{} = iter) do
    NIF.iter_seek_to_first(iter.ref)
  end

  def iter_seek_to_last(%Iterator{} = iter) do
    NIF.iter_seek_to_last(iter.ref)
  end

  def iter_key(%Iterator{} = iter) do
    NIF.iter_key(iter.ref)
  end

  def iter_value(%Iterator{} = iter) do
    NIF.iter_value(iter.ref)
  end

  def iter_next(%Iterator{} = iter) do
    NIF.iter_next(iter.ref)
  end

  def iter_prev(%Iterator{} = iter) do
    NIF.iter_prev(iter.ref)
  end

  def scan(db, cf, num \\ nil)
  def scan(%DB{}, %ColumnFamily{}, 0), do: []
  def scan(%DB{} = db, %ColumnFamily{} = cf, nil), do: scan(db, cf, -1)
  def scan(%DB{} = db, %ColumnFamily{} = cf, num) when is_integer(num) do
    NIF.scan(db.ref, cf.ref, num)
  end
end
